go 1.11 fails with crypto/x509 and cgo

## Issue description

With go 1.11 installed via nix on darwin, compiling simple commands that import crypto/x509 via net/http fail to compile due to problems with darwin core header imports.

Error message: 'CoreFoundation/CoreFoundation.h' file not found

For the record, go itself builds fine via nix. xcode command line tools seem to be installed ok, at least `xcode-select --install` claims they're installed already.

### Steps to reproduce

```
$ cat x.go
package main

import (
	"net/http"
)

func main() {
	print("hello")
}
$ go build -o x x.go
# crypto/x509
/nix/store/mkcsimqjfw9s2b9rn77yvx5bpfgavpp2-go-1.11.5/share/go/src/crypto/x509/root_cgo_darwin.go:16:10: fatal error: 'CoreFoundation/CoreFoundation.h' file not found
#include <CoreFoundation/CoreFoundation.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
$ CGO_ENABLED=0 go build -o x x.go
# command-line-arguments
./x.go:4:2: imported and not used: "net/http"
```

(changing x.go to make proper use of the http package results in a working binary)

## Technical details

 - system: `"x86_64-darwin"`
 - host os: `Darwin 18.2.0, macOS 10.14.3`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.1.3`
 - channels(rob): `"nixpkgs-19.03pre170546.0f43790d2ce"`
 - nixpkgs: `/Users/rob/.nix-defexpr/channels/nixpkgs`

I still get this error with go-1.12.6 on darwin.

Here's how I can produce the error. Create a file named `x.go`:

```go
package test

/*
#cgo CFLAGS: -mmacosx-version-min=10.6 -D__MAC_OS_X_VERSION_MAX_ALLOWED=1080
#cgo LDFLAGS: -framework CoreFoundation -framework Security

#include <CoreFoundation/CoreFoundation.h>
*/
import "C"
```
Build fails with go from nixpkgs, but works with `CC=clang`.
```
$ go build x.go
# command-line-arguments
./x.go:9:10: fatal error: 'CoreFoundation/CoreFoundation.h' file not found
#include <CoreFoundation/CoreFoundation.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
$ CC=clang go build x.go
$
```
I also get similar results calling go's CC directly:
```
$ cat xx.c
#include <CoreFoundation/CoreFoundation.h>
int main(){ return 0; }
$ go env | grep ^CC
CC="/nix/store/p9xdhw6viqrypbr09l7ng7g5065kpdm1-clang-wrapper-7.1.0/bin/cc"
$ /nix/store/p9xdhw6viqrypbr09l7ng7g5065kpdm1-clang-wrapper-7.1.0/bin/cc -framework CoreFoundation xx.c
xx.c:1:10: fatal error: 'CoreFoundation/CoreFoundation.h' file not found
#include <CoreFoundation/CoreFoundation.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
$ /usr/bin/clang -framework CoreFoundation xx.c
```

When trying to produce a minimal example it turned out that the problem I have is actually https://github.com/NixOS/nixpkgs/issues/69401, which has something to do with permissions on `$GOPATH`. So this issue does not apply - which is good, I guess.

For posterity: What I was really trying to do was using `github.com/stretchr/testify` in my tests, and that needs hashing somewhere in its dependencies. But for simplicity let us take @svend's minimal example:

`$ cat main.go`
```go
package main

/*
#cgo CFLAGS: -mmacosx-version-min=10.6 -D__MAC_OS_X_VERSION_MAX_ALLOWED=1080
#cgo LDFLAGS: -framework CoreFoundation -framework Security

#include <CoreFoundation/CoreFoundation.h>
*/
import "C"

func main() {}
```
`$ cat go.mod`
```go
module example.com/x
```
 `$ cat default.nix`
```nix
let
  nixpkgs = builtins.fetchGit {
    name = "nixpkgs";
    url = "https://github.com/NixOS/nixpkgs-channels";
    ref = "nixos-19.09";
    rev = "b9cb3b2fb2f45ac8f3a8f670c90739eb34207b0e";
  };
  pkgs = import (nixpkgs) {};
in
pkgs.buildGoModule {
  pname = "x";
  version = "0";
  modSha256 = "0sjjj9z1dhilhpc8pq4154czrb79z9cm044jvn75kxcjv6v5l2m5";
  src = ./.;
}
```

Then we try to build the thing inside `nix-shell`:
```shell
$ nix-shell
$ go build
go: failed to create cache directory /nix/store/3702mrbj6jc9vry6m9bdcl2qany7892r-go-1.12.16/share/go/pkg/mod/cache: mkdir /nix/store/3702mrbj6jc9vry6m9bdcl2qany7892r-go-1.12.16/share/go/pkg/mod: permission denied
```

This is what happens in the above issue. Let us work around it with sudo:

```shell
$ sudo go build
# example.com/x
./x.go:7:10: fatal error: 'CoreFoundation/CoreFoundation.h' file not found
#include <CoreFoundation/CoreFoundation.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

Here I got stuck yesterday, but then I remembered that `sudo` spawns a new environment. We need to carry over the one from `nix-shell` with `sudo -E`:

```shell
$ sudo -E go build
$ ls
default.nix  go.mod  x	x.go
```
So that worked. Unfortunately `go.mod` and the `x` binary are now owned by `root`. As noted in https://github.com/NixOS/nixpkgs/issues/69401#issue-498086212, unsetting `$GOPATH` fixes the initial permission problem. Let us fix this in `default.nix`:
 `$ cat default.nix`
```nix
# ...
pkgs.buildGoModule {
  # ...
  shellHook = ''
    unset GOPATH
  '';
}
```
Interestingly the issue does not occur when running `nix-build`, so this really is a problem with the Go environment, and has nothing to do with `buildGoModule`.

--- a/pkgs/development/compilers/go/1.13.nix
+++ b/pkgs/development/compilers/go/1.13.nix
@@ -186,8 +186,11 @@ stdenv.mkDerivation rec {
 
     export PATH=$(pwd)/bin:$PATH
 
+    ${optionalString (stdenv.buildPlatform != stdenv.targetPlatform) ''
     # Independent from host/target, CC should produce code for the building system.
+    # We only set it when cross-compiling.
     export CC=${buildPackages.stdenv.cc}/bin/cc
+    ''}
     ulimit -a
   '';
 

--- a/pkgs/development/compilers/go/1.14.nix
+++ b/pkgs/development/compilers/go/1.14.nix
@@ -193,8 +193,11 @@ stdenv.mkDerivation rec {
 
     export PATH=$(pwd)/bin:$PATH
 
+    ${optionalString (stdenv.buildPlatform != stdenv.targetPlatform) ''
     # Independent from host/target, CC should produce code for the building system.
+    # We only set it when cross-compiling.
     export CC=${buildPackages.stdenv.cc}/bin/cc
+    ''}
     ulimit -a
   '';
 

